src/datadog/remote_config/remote_config.cpp

```diff
@@ -168,12 +168,26 @@ nlohmann::json Manager::make_request_payload() {
 
 void Manager::process_response(const nlohmann::json& json) {
   state_.error_message = nullopt;
+  struct UpdateTargetsVersion {
+    decltype(state_)& state;
+    std::uint64_t new_targets_version;
+
+    ~UpdateTargetsVersion() noexcept {
+      // if there was a global error, the old targets_version should be sent in
+      // the next request (at least this is what the system-tests expect)
+      if ((!state.error_message || state.error_message->empty()) &&
+          new_targets_version != std::uint64_t(-1)) {
+        state.targets_version = new_targets_version;
+      }
+    }
+  } update_targets_version{state_, std::uint64_t(-1)};
```

Comment:
This seems a bit complex, especially when it might be simpler to just update the field after the entire process is successful. Am I missing something here?

Reply:
that's basically it. It's just that the function has multiple returns. But it can be done other ways of course (goto, new wrapping function, more nested ifs, etc.)

Reply:
Hey @cataphract, your solution is clever, but the RC code is already quite complex (which is entirely my fault) and not easy to grasp. Adding this defer behavior would create unnecessary additional cognitive load. I took the liberty of pushing a commit that I believe is simpler and easier to understand.